Add BamNative datatype

bgruening opened the pull request:
* add a new BamNative datatype
  * BamNative will not sort or index the Bam file
* Bam now subclasses from BamNative
* new test tool

bgruening committed:
add BamNative datatype

bgruening committed:
add new test file

bgruening committed:
add new test tool

bgruening committed:
add new bam_native datatype to the registry

bgruening committed:
bam to bigwig converter, converts also bam_native

bgruening committed:
use unsorted sam file for tests

bgruening committed:
add two converters

mvdbeek commented:
Can we make this queryname sorted instead/additionally ? If we can't assume reads are queryname-sorted (i.e as produced by aligners) many tools (duplicate marking, mapq correctors, split read extactors come to mind) would still need to either implement a check to see if the output is readname sorted or worse, sort them regardless of datatype.

bgruening commented:
Mh, this datatype was meant to not touch the BAM file, take it as it is produced and do not sort at all (and not index), hence the name.

If tools define `bam` as input the converter comes into play and convert `bam_native` files to `bam`, right?

mvdbeek commented:
I was assuming the point of this datatype is to avoid unnecessary conversions ?!
I.e `paired-end fastq` -> `bwa-mem` (which does coordinate sorting at the end) -> `samblaster` (which does queryname sorting for the input and coordinate sorting for the output). Instead we could directly go `paired-end fastq` -> `bwa-mem` -> `samblaster` and then, if necessary, coordinate sorting.

bgruening commented:
@mvdbeek it should address the issue in: https://github.com/galaxyproject/tools-iuc/pull/1590/files

`--reorder` will take the order as it is in the input files, so no guarantee that it is name sorted afaik.
But given this PR we probably can add a third type or add a metadata element that indicates the sorting.

mvdbeek commented:
> --reorder will take the order as it is in the input files, so no guarantee that it is name sorted afaik.

Right, which is the default for most aligners, I believe, but you're right in that this isn't strictly 'SO:queryname'. Maybe some additional metadata is the way to go that indicates that at least pairs of reads are adjacent ? Because that would be good enough for some important tools, but with the BamNative datatype as it is now this could be totally random. So basically only tools that absolutely do not care at all about sort order can consume this datatype, but I'm afraid people could start treating it as a queryname sorted file (because it could be queryname sorted).

bgruening commented:
@mvdbeek I just tried the following.
```
       <actions>
            <action name="sort_order" type="metadata" default="name_sorted_bag" />
       </actions>
```

With some hacks we could use this and skip/adjust the set_meta function in a way that the index is not created/used. We will probably miss some metadata, but this would enable us to use just one BAM file.
Disadvantage, is that we could still not upload name-sorted BAM files afaik.

ping also @blankenberg to gather more comments

mvdbeek commented:
> With some hacks we could use this and skip/adjust the set_meta function in a way that the index is not created/used. We will probably miss some metadata, but this would enable us to use just one BAM file.

Yeah, that does look like a good idea. If you know that a tool places reads with the same queryname next to each other (which isn't necessarily what SO:queryname implies) you could use the action to set the metadata accordingly. So in a first instance tool wrappers could handle this logic in the command section, and eventually we could have a syntax to request this format from the framework.

Maybe something like 
```
<param argument="--input" type="data" format="bam,sam" metadata_requirement="adjacent_queryname"/>
```
?

bgruening commented:
@mvdbeek what about the upload problem? Such "unordered" files could not be uploaded afaik.

Your example would then be already possible by using something like this I think:
```
<param argument="--input" type="data" format="bam,sam" >
     <options options_filter_attribute="metadata.sorted_by" >
          <filter type="add_value" value="adjacent_queryname" />
     </options>
     <validator type="expression" message="">value is not None and value.metadata.sorted == "adjacent_queryname"</validator>
</param>
```

blankenberg commented:
I like this approach. The default action is to still sniff then sort and index bam files when uploaded by a user, but they can manually upload an un/differently-sorted bam.

bgruening commented:
@blankenberg do you prefer this more explicit way or the alternative of having one BAM format with metadata indicating the sorting. The latter can probably not uploaded.

bgruening commented:
@blankenberg or to be more concrete, do you like this PR or the way I sketched out the metadata approach?

blankenberg commented:
I like this PR, and is what I would have done myself to support unsorted bam. I think the best thing we can probably do at this point is to have the 2+ datatypes. So we can upload un/diff-sorted bam for tools that will need them this way and also don't need to worry about the many existing tools that expect 'bam' to be sorted and indexed and any potential (silent) errors that may be caused.

The use of the metadata field values for limiting inputs can be useful and powerful, but I think they can be a bit clunky to use when developing tools especially in cases when things can be represented by the hierarchical structure of datatypes:
```
bam_native --> [ bam_sorted --> ] sortedBamA(e.g. current BAM),sortedBamB,sortedBamC
```
The hierarchy approach can fall apart a bit when the key/important attribute is greater than one in number. But if sort_order is really the primary defining metadata value, it makes sense to use multiple datatypes.

bgruening commented:
Ok, thats fine. Than this PR is ready for review. @mvdbeek any objections? Should I add a third datatype as well in this PR?

mvdbeek commented on lib/galaxy/datatypes/binary.py:
```diff
@@ -210,75 +207,19 @@ class Bam(Binary):
     def merge(split_files, output_file):
         """
-        Merges BAM files
+        Merges Bam files
```
Can we leave this at `BAM files` ? xref https://github.com/galaxyproject/galaxy/pull/5037#discussion_r155796306

bgruening replied:
Done.

mvdbeek commented on lib/galaxy/datatypes/binary.py:
```diff
@@ -374,6 +308,98 @@ def display_data(self, trans, dataset, preview=False, filename=None, to_ext=None
+            # fail metadata to end in the error state
+            pass
+
+    def sniff(self, filename):
```
I'd move that into the base class, and use `return super(Bam, self).sniff(filename) and self.dataset_content_needs_grooming(filename) ` (and maybe change `filename` to `file_name` for consistency ?)

bgruening replied:
Make sense, but it should be `return super(Bam, self).sniff(filename) and __not__ self.dataset_content_needs_grooming(filename)` right?

mvdbeek commented on test/functional/tools/sam_to_bam_native.xml:
```diff
@@ -0,0 +1,20 @@
+    ]]>
+    </command>
+    <inputs>
+        <param name="input1" type="data" format="bam" label="Concatenate Dataset"/>
```
Wouldn't this test use the `sam_to_bam` converter ?
Maybe change the input format to `bam_native` ?
If we add this to test/functional/tools/samples_tool_conf.xml it should be run during the functional tests.

bgruening replied:
I fixed this test, it should test the conversion from SAM to BAM (unsorted).
Thanks @mvdbeek for the review.

mvdbeek commented on lib/galaxy/datatypes/binary.py:
```diff
@@ -327,7 +257,11 @@ def to_archive(self, trans, dataset, name=""):
     def get_chunk(self, trans, dataset, offset=0, ck_size=None):
         index_file = dataset.metadata.bam_index
-        with pysam.AlignmentFile(dataset.file_name, "rb", index_filename=index_file.file_name) as bamfile:
+        if index_file:
```
We don't actually need the index at all for `get_chunk`, so we should probably drop lines 259-263.

bgruening committed:
Bam to BAM

bgruening committed:
rearrange sniffer

bgruening committed:
adjust tests

mvdbeek commented on test/functional/tools/sam_to_bam_native.xml:
```diff
@@ -0,0 +1,27 @@
+    ]]>
+    </command>
+    <inputs>
+        <param name="input" type="data" format="sam" label="SAM file"/>
```
This still doesn't test the converter, it just runs this tool. You should be able to specify as input format `bam_native` but provide a sam file in the test. Copy that to an output, and the output should be an unsorted bam file.

bgruening replied:
Ah, I see. No this test is testing something different. It tests that Galaxy is not sorting this file.
Do you want to see an additional test for the implicit converter as well?

mvdbeek committed:
Fix test and add datatypes to test/functional/tools/sample_datatypes_conf.xml

This test is still failing because `CONVERTER_sam_to_bam` will be invoked instead
of CONVERTER_sam_to_bam_native. Possibly becasue Bam inherits from BamNative, so
this isn't strictly speaking wrong, though not what one would expect.

mvdbeek commented:
@bgruening I added the test I was thinking about in https://github.com/bgruening/galaxy/pull/37.
This points out another issue, which is that the conversion to `bam_native` is going to be `bam` because `Bam` inherits from `BamNative`. I'm just not sure if this is really a problem. After all unsorted may be sorted.

mvdbeek committed:
Adjust sniffer test for unsorted bam

bgruening committed:
Merge pull request #37 from mvdbeek/bam_native_tests

Fix test and add datatypes to test/functional/tools/sample_datatypes_…

bgruening commented:
> the conversion to bam_native is going to be bam because Bam inherits from BamNative. 

What do you mean here? The conversion from Bam to BamNative?

> After all unsorted may be sorted.

That is true. A BamNative could be sorted as well. But I would not use the words sorted/unsorted here. Native should just refer to a BAM file that is not touched by Galaxy. Thats by the conversion from Bam to BamNative does not make much sense and has probably no use-case.

mvdbeek commented:
> Thats by the conversion from Bam to BamNative does not make much sense and has probably no use-case.

Right, but if you use a `sam` dataset to a `bam_native` input, the `sam_to_bam` converter will be used, instead of `sam_to_bam_native`. The `sam_to_bam_native` will actually not be used for implicit conversions right now. But maybe that's fine.

bgruening commented:
Interesting this seems to be a (unrelated to this PR) bug. If you use this instead:

```
    <datatype extension="sam" type="galaxy.datatypes.tabular:Sam" display_in_upload="true">
        <converter file="sam_to_bam_native.xml" target_datatype="bam_native"/>
        <converter file="sam_to_bam.xml" target_datatype="bam"/>
        <converter file="sam_to_bigwig_converter.xml" target_datatype="bigwig"/>
    </datatype>
```

tt works and sam_to_bam_native is used. So Galaxy seems to take the first in the list and not matches target_datatype.

mvdbeek commented:
I guess it's not too bad of a mechanism to use the first, given that both are valid targets. If you order it that way and the tests pass that would be good enough for me.

bgruening committed:
resort the converters so that the accepted_formats are infront

bgruening commented:
> I guess it's not too bad of a mechanism to use the first, given that both are valid targets. If you order it that way and the tests pass that would be good enough for me.

This would mean that all implicit converters would favor the bam_native conversion.

Please have a look at this PR: https://github.com/bgruening/galaxy/pull/38
This implements at least the behavior I would expect. If the tool wants to have BAM you see `(as bam)` if the tool wants to have BamNative you see now `(as bam_native)`.

This is not implemented to be performant atm, ideally this should be moved into the caching a function up. ideas welcome.

mvdbeek commented:
> This would mean that all implicit converters would favor the bam_native conversion.

I think that would only happen if `bam_native` and `bam` are valid. Which is `x` to `bam_native` at the moment. `bam` is more restricted, so if  a tool demands `bam` it will get `bam`.

But I can see how it makes sense to prefer an exact match, I'd implement that at `get_converters_by_datatype` though, which is already being cached.

bgruening committed:
implement priority converters

bgruening committed:
fix spelling

bgruening committed:
Merge pull request #38 from bgruening/converter_resorting

resort the converters so that the accepted_formats are infront

mvdbeek committed:
Keep header when converting sam to bam

Otherwise the converted datasets are not valid BAM files.

mvdbeek committed:
Revert "fix spelling"

This reverts commit b139726ed380192ecf03f10e91220a0488b573f3.

mvdbeek committed:
Revert "implement priority converters"

This reverts commit 8928b34506c1ba1f4856fad82dc42aef0f7c68f6.

mvdbeek committed:
Move up sam_to_bam_native converter

That should allow implicit conversion from sam to bam_native for tools that
support bam_native.

mvdbeek committed:
Revert "resort the converters so that the accepted_formats are infront"

This reverts commit 446643159540f8db67dfd0752560b387b63fc4b8.

mvdbeek committed:
Enhance tests for sam_to_bam conversion

This demonstrates that the `sam_to_bam` converter will be used for an input
with `format="bam"`, and not `sam_to_bam_native`.

bgruening commented:
Meh, this looks like a real failing test ...

mvdbeek commented:
Yeah, I got some local commits for this

mvdbeek commented:
With https://github.com/bgruening/galaxy/pull/39 the tests should pass, but we'll need to rebase after the merge. I think we can and should do this without the additional sorting for now. 

Otherwise we'd probably have to refine the caching of the results, because the first time we ask for conversions we store the order. Something like `functools.lru_cache` instead of caching the result in `self._converters_by_datatype[ext]` would work, but I'm not convinced we need the added complexity.

bgruening commented:
@mvdbeek I think I found the difference, try this:

```
        <converter file="sam_to_bam.xml" target_datatype="bam"/>
        <converter file="sam_to_bam_native.xml" target_datatype="bam_native"/>
        <converter file="sam_to_bigwig_converter.xml" target_datatype="bigwig"/>
```
With your PR this will not work. Can you confirm?

mvdbeek commented:
Right, but why would you do this ?
The datatypes in general are sensitive to the order in which they are defined / listed in the datatypes_conf.xml, so at least that is consistent.

bgruening commented:
> The datatypes in general are sensitive to the order in which they are defined / listed in the datatypes_conf.xml, so at least that is consistent.

afaik only the sniffers are sensible to the order not the datatypes. I think it is odd if such a functionality is depended on the order if it does not need to. But its your call. Happy to merge it.

mvdbeek commented:
I'd be more comfortable not changing the way the converters are picked right now. I agree that if you ask for an extension (`bam_native` in this case) you should get that extension and not the extension of a subclass. I'd prefer to work on this in a separate PR if that doesn't bother you too much.

bgruening committed:
Merge pull request #39 from mvdbeek/bam_unsorted

Fix tests and revert additional sorting of converters

bgruening committed:
Merge branch 'dev' into bam_unsorted

bgruening commented:
Sure, if you prefer to keep it separate. I have merged your PR and resolved the conflict hopefully.

mvdbeek commented:
Hmm, those failing tests are passing when I test them locally. I wonder if this is because Jenkins runs an older version of samtools ?

mvdbeek commented:
Arg, I was thinking we could get around the samtools version problem by using `pysam.view`, but the implementation is only partial and doesn't support writing to a file. So what now ? Upgrading samtools in the testing container would be my next suggestion ...

nsoranzo commented:
Maybe the failing framework tests can be fixed by auto-installing the samtools requirements with `conda_auto_install = True` option in `config/galaxy.ini`?

mvdbeek committed:
Enable conda_auto_init and conda_auto_install for framework tests

mvdbeek commented:
We should be eating our own dogfood, right ? :D

martenson commented:
I wonder how much slower it will make the tests.

mvdbeek commented:
Shouldn't be much slower, around a minute or so.

On Jan 17, 2018 5:18 PM, "Martin Cech" <notifications@github.com> wrote:

> I wonder how much slower it will make the tests.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/galaxyproject/galaxy/pull/5180#issuecomment-358356272>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGfVpVs6SQx-ZHBhVVo6Wh1IQygFyYr4ks5tLh0KgaJpZM4Q8gmy>
> .
>

mvdbeek committed:
Fix datatype order

nsoranzo commented on test/base/driver_util.py:
```diff
@@ -851,6 +856,8 @@ def setup(self, config_object=None):
                     prefer_template_database=getattr(config_object, "prefer_template_database", False),
                     log_format=log_format,
+                    conda_auto_init=self.conda_auto_init,
+                    conda_auto_install=self.conda_auto_install,
```
Should these have value `getattr(config_object, 'conda_auto_init', False)` and `getattr(config_object, 'conda_auto_install', False)` respectively?
If yes, then probably the new lines 799-800 are not needed.

mvdbeek replied:
Yes, that should work!

mvdbeek committed:
Simplify conda framework setup

Thanks @nsoranzo!

bgruening commented:
Cool, thanks @mvdbeek seems to work!

mvdbeek reviewed (approved):
Looks good to me! Keep in mind that BamNative makes no guarantees to sort oder at all. 

We should (in another PR) implement some variants of this for the cases where reads are in the order that has been spit out by the aligner (`BamReadsAdjacent` ?) and where reads are ordered by queryname (`BamQueryNameSorted` ?) and by tag (`BamTagSored` ?).

mvdbeek commented:
Thanks a lot @bgruening !

bgruening commented:
Thanks to you @mvdbeek!
Do you want me to get this ordering of datatypes in or not?

martenson commented:
>We should (in another PR) implement some variants of this for the cases where reads are in the order that has been spit out by the aligner (BamReadsAdjacent ?) and where reads are ordered by queryname (BamQueryNameSorted ?) and by tag (BamTagSored ?).

This warranties an issue I think
